runtime/src/main/java/org/corfudb/runtime/collections/SMRMultiIndex.java

```diff
@@ -0,0 +1,362 @@
+     * and the parameters to the remove call.
+     *
+     * @param index             The previous state of the map
+     * @param key               The key from the remove call
+     * @return                  An undo record, which for a remove is the
+     *                          previous value in the map.
+     */
+    @DontInstrument
+    V undoRemoveRecord(SMRMultiIndex<K,V,I,P> index, K key) {
+        return index.get(key);
+    }
+
+    /** Undo a remove, given the current state of the map, an undo record
+     * and the arguments to the remove command to undo.
+     *
+     * @param index           The state of the map after the put to undo
+     * @param undoRecord    The undo record generated by undoRemoveRecord
+     */
+    @DontInstrument
+    void undoRemove(SMRMultiIndex<K,V,I,P> index, V undoRecord, K key) {
+        if (undoRecord == null) {
+            index.remove(key);
+        } else {
+            index.put(key, undoRecord);
+        }
+    }
+
+    @Mutator(name = "putAll")
+    @Override
+    public void putAll(Map<? extends K, ? extends V> m) {
+
+        throw new RuntimeException("###########Operation Not Implemented.##########");
+    }
+
+
+    /** Get a value by it's row key.
+     *
+     * @param key    The row key to retrieve.
+     * @return          The object at the row key, or
+     *                  NULL, if no object was mapped.
+     */
+    @Accessor
+    @Nullable
+    public V getByRowIndex(K key) {
+        return mainMap.get(key);
+    }
+
+
+    /** Get a value by a given named index.
+     *
+     * @param indexName  The index of the index function from the list
+     *                  passed in at construction time.
+     * @param indexKey The index Key to use.
+     * @return          A set of values which match the given index criteria.
+     */
+    @Accessor
+    public Collection<Map.Entry<K,P>> getByNamedIndex(String indexName, I indexKey) {
+        Map<I,Map<K,P>> indexMap = indexMaps.get(indexName);
+        if (indexMap == null)
+            return Collections.EMPTY_SET;
+        Map<K,P> index = indexMap.get(indexKey);
+        if (index == null)
+            return Collections.EMPTY_SET;
+        return index.entrySet();
+    }
+
+    /** Get the size of the multi-index.
+     * @return  The number of entries mapped (rows).
+     */
+    @Accessor
+    public int size() {
+        return mainMap.size();
+    }
+
+    @Accessor
+    @Override
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    @Accessor
+    @Override
+    public boolean containsKey(Object key) {
+        return mainMap.containsKey(key);
+    }
+
+    @Accessor
+    @Override
+    public boolean containsValue(Object value) {
+        return mainMap.containsValue(value);
+    }
+
+
+
+    /** Clear the multi-index.
+     */
+    @Mutator(reset = true, name = "clear")
+    public void clear() {
+        throw new RuntimeException("###########Operation Not Implemented.##########");
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00112)